File: akka-docs-new/src/test/scala/tutorial_3/DeviceGroup.scala
/**
 * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
 */
package tutorial_3

import akka.actor.{ Actor, ActorLogging, ActorRef, Cancellable, Props, Stash, Terminated }
import tutorial_3.DeviceGroup._
import tutorial_3.DeviceManager.RequestTrackDevice
import scala.concurrent.duration._

object DeviceGroup {

  def props(groupId: String): Props = Props(new DeviceGroup(groupId))

  case class RequestDeviceList(requestId: Long)
  case class ReplyDeviceList(requestId: Long, ids: Set[String])

  case class RequestAllTemperatures(requestId: Long)
  case class RespondAllTemperatures(requestId: Long, temperatures: Map[String, Option[Double]])

  case class CollectionTimeout(requestId: Long)
}

class DeviceGroup(groupId: String) extends Actor with ActorLogging with Stash {
  var deviceIdToActor = Map.empty[String, ActorRef]
  var actorToDeviceId = Map.empty[ActorRef, String]
  var nextCollectionId = 0L

  override def preStart(): Unit = log.info(s"DeviceGroup $groupId started")

  override def postStop(): Unit = log.info(s"DeviceGroup $groupId stopped")

  override def receive: Receive = waitingForRequest orElse generalManagement

  def waitingForRequest: Receive = {
    case RequestAllTemperatures(requestId) =>
      import context.dispatcher

      val collectionId = nextCollectionId
      val requester = sender()
      nextCollectionId += 1
      val answersSoFar = deviceIdToActor.mapValues(_ => None)
      context.children.foreach(_ ! Device.ReadTemperature(collectionId))
      val collectionTimeoutTimer = context.system.scheduler.scheduleOnce(3.seconds, self, CollectionTimeout(collectionId))

      context.become(
        collectResults(
          collectionTimeoutTimer,
          collectionId,
          requester,
          requestId,
          answersSoFar.size,
          answersSoFar) orElse generalManagement,
        discardOld = false
      )
  }

  def generalManagement: Receive = {
    // Note the backticks
    case trackMsg @ RequestTrackDevice(`groupId`, _) =>
      handleTrackMessage(trackMsg)

    case RequestTrackDevice(groupId, deviceId) =>
      log.warning(s"Ignoring TrackDevice request for $groupId. This actor is responsible for ${this.groupId}.")

    case RequestDeviceList(requestId) =>
      sender() ! ReplyDeviceList(requestId, deviceIdToActor.keySet)

    case Terminated(deviceActor) =>
      removeDeviceActor(deviceActor)
  }

  def handleTrackMessage(trackMsg: RequestTrackDevice): Unit = {
    deviceIdToActor.get(trackMsg.deviceId) match {
      case Some(ref) =>
        ref forward trackMsg
      case None =>
        log.info(s"Creating device actor for ${trackMsg.deviceId}")
        val deviceActor = context.actorOf(Device.props(groupId, trackMsg.deviceId), "device-" + trackMsg.deviceId)
        context.watch(deviceActor)
        deviceActor forward trackMsg
        deviceIdToActor += trackMsg.deviceId -> deviceActor
        actorToDeviceId += deviceActor -> trackMsg.deviceId
    }
  }

  def removeDeviceActor(deviceActor: ActorRef): Unit = {
    val deviceId = actorToDeviceId(deviceActor)
    log.info(s"Device actor for $deviceId has been terminated")
    actorToDeviceId -= deviceActor
    deviceIdToActor -= deviceId
  }

  def collectResults(
    timer:        Cancellable,
    expectedId:   Long,
    requester:    ActorRef,
    requestId:    Long,
    waiting:      Int,
    answersSoFar: Map[String, Option[Double]]
  ): Receive = {

    case Device.RespondTemperature(`expectedId`, temperatureOption) =>
      val deviceActor = sender()
      val deviceId = actorToDeviceId(deviceActor)
      val newAnswers = answersSoFar + (deviceId -> temperatureOption)

      if (waiting == 1) {
        requester ! RespondAllTemperatures(requestId, newAnswers)
        finishCollection(timer)
      } else {
        context.become(collectResults(timer, expectedId, requester, requestId, waiting - 1, newAnswers))
      }

    case Terminated(deviceActor) =>
      val deviceId = actorToDeviceId(deviceActor)
      removeDeviceActor(deviceActor)
      val newAnswers = answersSoFar + (deviceId -> None)

      if (waiting == 1) {
        requester ! RespondAllTemperatures(requestId, newAnswers)
        finishCollection(timer: Cancellable)
      } else {
        context.become(collectResults(timer, expectedId, requester, requestId, waiting - 1, newAnswers))
      }

    case CollectionTimeout(`expectedId`) =>
      requester ! RespondAllTemperatures(requestId, answersSoFar)
  }

  def finishCollection(timer: Cancellable): Unit = {
    context.unbecome()
    timer.cancel()
  }

}

File: akka-docs-new/src/test/java/jdocs/tutorial_3/DeviceGroup.java
/**
 * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
 */
package jdocs.tutorial_3;

import java.util.Set;
import java.util.Map;
import java.util.HashMap;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.Terminated;
import akka.event.Logging;
import akka.event.LoggingAdapter;

import jdocs.tutorial_3.Device;
import jdocs.tutorial_3.DeviceManager;

//#device-group-full
public class DeviceGroup extends AbstractActor {
  private final LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);

  final String groupId;

  public DeviceGroup(String groupId) {
    this.groupId = groupId;
  }

  //#device-group-register
  public static Props props(String groupId) {
    return Props.create(DeviceGroup.class, groupId);
  }
  //#device-group-register

  public static final class RequestDeviceList {
    final long requestId;

    public RequestDeviceList(long requestId) {
      this.requestId = requestId;
    }
  }

  public static final class ReplyDeviceList {
    final long requestId;
    final Set<String> ids;

    public ReplyDeviceList(long requestId, Set<String> ids) {
      this.requestId = requestId;
      this.ids = ids;
    }
  }
  //#device-group-register
//#device-group-register
//#device-group-register
//#device-group-remove

  Map<String, ActorRef> deviceIdToActor = new HashMap<>();
  //#device-group-register
  Map<ActorRef, String> actorToDeviceId = new HashMap<>();
  //#device-group-register

  @Override
  public void preStart() {
    log.info("DeviceGroup {} started", groupId);
  }

  @Override
  public void postStop() {
    log.info("DeviceGroup {} stopped", groupId);
  }

  @Override
  public Receive createReceive() {
    return receiveBuilder()
            .match(DeviceManager.RequestTrackDevice.class, trackMsg -> {
              if (this.groupId.equals(trackMsg.groupId)) {
                ActorRef deviceActor = deviceIdToActor.get(trackMsg.deviceId);
                if (deviceActor != null) {
                  deviceActor.forward(trackMsg, getContext());
                } else {
                  log.info("Creating device actor for {}", trackMsg.deviceId);
                  deviceActor = getContext().actorOf(Device.props(groupId, trackMsg.deviceId), "device-" + trackMsg.deviceId);
                  //#device-group-register
                  getContext().watch(deviceActor);
                  actorToDeviceId.put(deviceActor, trackMsg.deviceId);
                  //#device-group-register
                  deviceIdToActor.put(trackMsg.deviceId, deviceActor);
                  deviceActor.forward(trackMsg, getContext());
                }
              } else {
                log.warning(
                        "Ignoring TrackDevice request for {}. This actor is responsible for {}.",
                        groupId, this.groupId
                );
              }
            })
            .match(RequestDeviceList.class, r -> {
              getSender().tell(new ReplyDeviceList(r.requestId, deviceIdToActor.keySet()), getSelf());
            })
            .match(Terminated.class, t -> {
              ActorRef deviceActor = t.getActor();
              String deviceId = actorToDeviceId.get(deviceActor);
              log.info("Device actor for {} has been terminated", deviceId);
              actorToDeviceId.remove(deviceActor);
              deviceIdToActor.remove(deviceId);
            })
            .build();
  }
}
//#device-group-remove
//#device-group-register
//#device-group-full
